### ISSUE TITLE
Crash when saving the token on disk with `AttributeError: DbfilenameShelf instance has no attribute '__exit__'`

### ISSUE THREAD

**Issue by MajinLogan** [pawelad/pymonzo#1]
When creating MonzoAPI object using environmental variables, I get the following error:

```
"AttributeError: DbfilenameShelf instance has no attribute '__exit__' "
```

This occurs on any method that is used to interact with the access token on disk.

**Comment by pawelad** [comment]
Could you please provide the whole stacktrace and the code that resulted in this error? Thanks.

**Comment by MajinLogan** [comment]
```
Traceback (most recent call last):
File "<stdin>", line 1, in <module>
File "build/bdist.linux-armv61/egg/pymonzo/monzo_api.py", line 97, in __init__
File "build/bdist.linux-armv61/egg/pymonzo/monzo_api.py", line 149, in _get_oauth_token
File "build/bdist.linux-armv61/egg/pymonzo/monzo_api.py", line 123, in _save_token_on_disk
AttributeError: DbfilenameShelf instance has no attribute '__exit__'
```

**Comment by MajinLogan** [comment]
It could be that I am running an older version of the shelve library?

**Comment by pawelad** [comment]
And what was the actual code that resulted in this? Was it just `MonzoAPI` object initialization?

To be honest I have no idea why that happens and I can't reproduce it. What distro (OS?) and Python version is this?

> It could be that I am running an older version of the shelve library?

`shelve` is in Python standard library so I don't think that should be a problem.

**Comment by MajinLogan** [comment]
Yes, this happens during object initialisation.

Out of interest, why are you using shelve and not just saving the token string in a text file using open() or close()? Or keeping the token string in the object?

I am running this on Rasbian Jessie, and have tried Python 2.7. also tried 3.4 but had a separate error come up for that.

On 1 Aug 2017 12:32, Paweł Adamczak <notifications@github.com> wrote:

And what was the actual code that resulted in this? Was it just MonzoAPI object initialization?

To be honest I have no idea why that happens and I can't reproduce it. What distro (OS?) and Python version is this?

It could be that I am running an older version of the shelve library?

shelve is in Python standard library so I don't think that should be a problem.

**Comment by bartonp** [comment]
It happens on Python 2.7 on both Windows 10, and Mac OS (whatever the latest version is).

This what I did to fix it for those versions at least -
 https://github.com/bartonp/pymonzo/commit/95e2edd2324d6a1bcc1f00fc615d28cd56557582?diff=unified - and you don't need the import json. That was there for testing purposes.

**Comment by pawelad** [comment]
@bartonp thanks for the confirmation and the fix, I'll try to find some time this weekend to take a look at it.

**Comment by pawelad** [comment]
OK, I just replicated this as well on Python 2.7.13 on OS X. Sorry it took so long but it's been a crazy couple of weeks.

I'll fix it in the next couple of days and make sure that the test suite covers this as well.

@MajinLogan so the quick fix for now (until I release a fix) is either to install the version patched by @bartonp or use Python 3

**Comment by pawelad** [comment]
> I am running this on Rasbian Jessie, and have tried Python 2.7. also tried 3.4 but had a separate error come up for that.

@MajinLogan what error did you get on Python 3.4? Could you open a new issue about it with the stacktrace? I have some free time and would love to squash these bugs

**Comment by delabere** [comment]
This is the error that I get with python 3 (which I have raised as a separate issue:
```
m.MonzoAPI()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/pymonzo/monzo_api.py", line 95, in __init__
    self._token = self._get_oauth_token()
  File "/usr/local/lib/python3.6/site-packages/pymonzo/monzo_api.py", line 144, in _get_oauth_token
    client_secret=self._client_secret,
  File "/usr/local/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 244, in fetch_token
    self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/usr/local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 409, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/usr/local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 376, in parse_token_response
    validate_token_parameters(params)
  File "/usr/local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 383, in validate_token_parameters
    raise_from_error(params.get('error'), params)
  File "/usr/local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/errors.py", line 404, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.ServerError: (server_error) Unknown error occurred
>>>
```

**Comment by pawelad** [comment]
Okay, so this issue should be resolved by the just released [v0.9.0](https://github.com/pawelad/pymonzo/releases/tag/v0.9.0)
